cpp/Naive.cpp

```diff
@@ -1,65 +1,413 @@
+                                                                    bracketAParUEKPar_K_Loop(kx, ky));
+                    // TODO(OPT) reuse star
+                    momentsNew_K(N_E, kx, ky) = exp_nu(kx, ky, hyper.nu_2, dt) * moments_K(N_E, kx, ky) +
+                                                dt / 2.0 * (1 + exp_nu(kx, ky, hyper.nu_2, dt)) * GM_Nonlinear_K(N_E, kx, ky) +
+                                                dt / 2.0 * GM_Nonlinear_K_Loop(N_E, kx, ky);
+
+                    phi_K_New(kx, ky) = nonlinear::phi(momentsNew_K(N_E, kx, ky), kx, ky);
+                });
+
+                derivatives(phi_K_New, dPhi_Loop);
+                DerivateNewMoment(N_E);
+
+                // Compute G2
+                auto bracketPhiG2_K_Loop = halfBracket(dPhi_Loop, sliceXY(dGM_Loop, G_MIN));
+                auto bracketAParG3_K_Loop = halfBracket(sliceXY(dGM_Loop, A_PAR), sliceXY(dGM_Loop, G_MIN + 1));
+
+                for_each_kxky([&](Dim kx, Dim ky) {
+                    GM_Nonlinear_K_Loop(G_MIN, kx, ky) = nonlinear::G2(bracketPhiG2_K_Loop(kx, ky),
+                                                                       bracketAParG3_K_Loop(kx, ky),
+                                                                       bracketAParUEKPar_K_Loop(kx, ky));
+                    // TODO(OPT) reuse star
+                    momentsNew_K(G_MIN, kx, ky) = exp_nu(kx, ky, hyper.nu_2, dt) * moments_K(G_MIN, kx, ky) +
+                                                  dt / 2.0 * (1 + exp_nu(kx, ky, hyper.nu_2, dt)) *
+                                                  GM_Nonlinear_K(G_MIN, kx, ky) +
+                                                  dt / 2.0 * GM_Nonlinear_K_Loop(G_MIN, kx, ky);
+                });
+                DerivateNewMoment(G_MIN);
+
+                DxDy<fftw::mdbuffer<2u>> dGMinusPlus_Loop{X, Y};
+                for (int m = G_MIN + 1; m < LAST; ++m) {
+                    for_each_xy([&](Dim x, Dim y) {
+                        dGMinusPlus_Loop.DX(x, y) =
+                                std::sqrt(m) * dGM_Loop.DX(m - 1, x, y) + std::sqrt(m + 1) * dGM_Loop.DX(m + 1, x, y);
+                        dGMinusPlus_Loop.DY(x, y) =
+                                std::sqrt(m) * dGM_Loop.DY(m - 1, x, y) + std::sqrt(m + 1) * dGM_Loop.DY(m + 1, x, y);
+                    });
+
+                    auto bracketAParGMMinusPlus_K_Loop = halfBracket(sliceXY(dGM_Loop, A_PAR), dGMinusPlus_Loop);
+                    auto bracketPhiGM_K_Loop = halfBracket(dPhi_Loop, sliceXY(dGM_Loop, m));
+
+                    for_each_kxky([&](Dim kx, Dim ky) {
+                        GM_Nonlinear_K_Loop(m, kx, ky) = nonlinear::GM(m, bracketPhiGM_K_Loop(kx, ky),
+                                                                       bracketAParGMMinusPlus_K_Loop(kx, ky));
+                        // TODO(OPT) reuse star
+                        momentsNew_K(m, kx, ky) = exp_gm(m, hyper.nu_ei, dt) * moments_K(m, kx, ky) +
```

Comment:
This update not quite right. Compare to REGK.jl ln 472-475

Reply:
Great catch, thanks!